[ARM] Add scalar add_sat costs.

davemgreen opened the pull request:
These can usually generate:
 - qadd / qsub for signed i32 scalars
 - uqadd16 / qadd16 / uqsub16 / qsub16 with an extend for signed/unsigned i8/i16
 - Are expanded to an add + cmp + sel otherwise

This can lead to differences in unrolling etc, but should be a better cost for the instructions.

arsenm commented on llvm/lib/Target/ARM/ARMTargetTransformInfo.cpp:
```diff
@@ -1904,17 +1905,37 @@ ARMTTIImpl::getIntrinsicInstrCost(const IntrinsicCostAttributes &ICA,
+    if (RetTy->isIntegerTy()) {
+      if (IsSigned && ST->hasDSP() && RetTy->isIntegerTy(32))
+        return 1; // qadd / qsub
+      if (ST->hasDSP() && (RetTy->isIntegerTy(8) || RetTy->isIntegerTy(16)))
```
Maybe check size <= 16? I assume other types would legalize up to 16-bit?

arsenm commented on llvm/lib/Target/ARM/ARMTargetTransformInfo.cpp:
```diff
@@ -1904,17 +1905,37 @@ ARMTTIImpl::getIntrinsicInstrCost(const IntrinsicCostAttributes &ICA,
+    bool IsSigned = (Opc == Intrinsic::sadd_sat || Opc == Intrinsic::ssub_sat);
+    Type *RetTy = ICA.getReturnType();
+
+    if (RetTy->isIntegerTy()) {
```
dyn_cast and check getBitWidth below?

davemgreen committed:
[ARM] Add scalar add_sat costs.

These can usually generate:
 - qadd / qsub for signed i32 scalars
 - uqadd16 / qadd16 / uqsub16 / qsub16 with an extend for signed/unsigned
   i8/i16
 - Are expanded to an add + cmp + sel otherwise

davemgreen commented:
Rebase and ping - thanks. Unfortunately these more accurate costs might cause some performance regressions according to my tests due to extra unrolling, but the better numbers should be better in general.
